### DISCUSSION TITLE
Private properties and Class Design

### DISCUSSION THREAD

**Discussion by penhorwood** [isar/isar#732] [Ideas]
Isar looks like a great database platform but I wonder why I need to choose to have public properties instead of private ones to save my data with isar.  The data isn't any different.  You could simply include an annotation to make it work.  Now my choices at a design level are being forced on me by the database options I choose.

**Comment by jannikrh** [comment]
You always have the possibility to design your models freely, if you distinguish and map between the Isar Collection models in the Data Layer and the Business models in the Domain Layer of your App.

**Comment by penhorwood** [comment]
I don't understand your comments.  I am super new to Isar so maybe I am missing something.  When I designed my class that I am using for storage on the Public properties are stored on the device.  Any private properties are not.  That limits my design of the class.  Here is an example.  In this example on the publicData point is stored not the _privateData.  The example is a bit contrived as there would be no need to store the privateData in this case.

```
import 'package:isar/isar.dart';
part 'storeme.g.dart';

@collection
class StoreMe {
	Id id;

	String publicData = 'new';
	String _privateData = 'really-private';

	StoreMe(): id = Isar.autoIncrement;
}```

**Comment by penhorwood** [comment]
Then there is this example.  This generates a "**Constructor parameter does not match a property.**" error.  You might not like that design but it is possible in Dart.

```
import 'package:isar/isar.dart';
part 'storeme.g.dart';

@collection
class StoreMe {
	Id id;

	String publicData;
	String _privateData;

	StoreMe({
		  required String data1
		, required String data2
	}) : id = Isar.autoIncrement
		, publicData = data1
		, _privateData = data2
	;
}
```

**Comment by jannikrh** [comment]
Of course, it is only one of many possibilities and points of view, but this one has helped me often.

The example class in your case `StoreMe` exists several times in the project. 
- There is once the Isar Collection in the Data Layer of the App. This has only public properties and other necessary characteristics required by Isar.
- And there is your actual business object `StoreMeBusinessModel`(the name could be the same), which you use in the app and where you have all the freedom to add private properties as well. 

In the business domain the object is then mapped. This makes your actual app independent of the database and you can theoretically replace Isar with another database without having to customize the whole app.

UI Layer - Business Layer (`StoreMeBusinessModel`) - Data Layer (`StoreMe`)

**Comment by penhorwood** [comment]
@jannikrh The case you present is a very normal use case in development.  I see if often.

If I choose to use a more direct approach to my data objects Isar limits my design.  I simple annotation would change that and make Isar work for more use cases.

**Reply by simc** [reply]
I also don't really see the need to persist private properties. Could you please explain why that is needed?

Could you please explain what kind of "simple annotation" would fix your issue?

**Comment by penhorwood** [comment]
The use case that comes to mind StateNotifier classes.  From their documentation.

// The StateNotifier class that will be passed to our StateNotifierProvider.
// This class should not expose state outside of its "state" property, which means
// **no public getters/properties!**
// The public methods on this class will be what allow the UI to modify the state.

Other use cases are where people come from other languages where using private data points is normal, like Java.  People want to persist their classes for some reason.  Again directly form java documentation:  **Persistent classes are those java classes whose objects have to be stored in the database tables.**

Another possible uses case.  Below is a possible class that I made up on the fly.
```
import 'package:isar/isar.dart';
part 'storeme.g.dart';

@collection
class Person {
	Id id;

	@include
	String _firstname;

	@include
	String _lastname;

	@include
	int _age = 0;

	String get first => _firstname;
	String get surname => _lastname;

	Person({
		  required String first
		, required String surname
	}): id = Isar.autoIncrement
		, _firstname = first
		, _lastname = surname
	;

	void setAge( DateTime DOB ){
		Duration diff = DateTime.now().difference(DOB);
		_age = diff.inDays;
	}
}
```

**Comment by penhorwood** [comment]
As for the annotation that might be more complicated the more I think about it.  It will depend on how the objects are actually stored and recreated from the database.  I didn't review the code so I know nothing about how that works.  The annotation could be simple the implementation of that annotation not so much.

**Comment by Mayb3Nots** [comment]
@leisim look at the code below. Here i wanna store a profile image however because `Uint8List` is not supported I have to convert the type from `List<int>` to `Uint8List` this is fine to convert to List however to convert from list I have to write this especially when is nullable
`_profileImage == null ? null : Uint8List.fromList(_profileImage!)`. Since I don't wanna write this everytime I need to get the profile image I created a getter in the class and store the `List<int>` in a private field. I could use different name for the saved field and getter but that could be error prone. Thats why private fields would be nice. Unless you have another way of doing this?


```dart

class FollowersNotification extends Equatable {

  FollowersNotification({
    required this.id,
    List<int>? profileImage,
    required this.createdOn,
    required this.username,
    required this.approved,
    required this.declined,
    required this.userID,
  }) : _profileImage = profileImage;

  factory FollowersNotification.fromFirestore(
    DocumentSnapshot data,
    UserDataImage userData,
  ) {
    final timeStamp = data['requested_on'] as Timestamp;
    final username = userData.data.username;
    if (username == null) {
      throw FirestoreMissingFieldValueException(
        'Missing user name from user data',
        ref: 'users/${userData.data.uid}',
        missingFieldName: 'username',
      );
    }
    return FollowersNotification(
      id: data.id.hashCode,
      profileImage: userData.image,
      createdOn: DateTime.fromMillisecondsSinceEpoch(
        timeStamp.millisecondsSinceEpoch,
      ),
      username: username,
      approved: data['approved'],
      declined: data['declined'],
      userID: data.id,
    );
  }
  final Id id;
  List<byte>? _profileImage;
  final DateTime createdOn;
  String username;
  bool approved;
  bool declined;
  final String userID;
  
  @ignore
  Uint8List? get profileImage =>
      _profileImage == null ? null : Uint8List.fromList(_profileImage!);
  @ignore
  @override
  List<Object?> get props => [id, userID, username];
}
```
